### ISSUE TITLE
Implement custom sorting

### ISSUE THREAD

**Issue by stacksjb** [alanvardy/tod#804]
See https://github.com/alanvardy/tod/issues/793

Implement sorting based on appropriate predefined parameters and methods:

due datetime
due date  (allow custom due time)
priority
labels
project
alphabetical
no sort (custom Todoist sort - default)

Allow layering of methods, such as priority then datetime

**Comment by stacksjb** [comment]
@alanvardy is there a way to stage test data (tasks) into tod for unit tests?

i.e. I'd like to load in 8 tasks with varying priorities, orders, and due dates (probably calculated as offset from date test is run) and then check they are sorted correctly for varying sort tests.

I get basic 'assert equals' unit tests but not sure about the test data part.

**Comment by alanvardy** [comment]
yes! Take a look at [test::fixtures](https://github.com/alanvardy/tod/blob/main/src/test.rs), and where I use it in the test suite, i.e. [here](https://github.com/alanvardy/tod/blob/310fd741205828556f4ae55b6be86af32d433145/src/tasks.rs#L753)

This should hopefully reduce the number of lines you need to write to create a task in test.

**Comment by alanvardy** [comment]
Currently these are customizable

```rust
pub struct SortValue {
    /// Task has one of these priorities
    pub priority_none: u8,
    pub priority_low: u8,
    pub priority_medium: u8,
    pub priority_high: u8,
    pub no_due_date: u8,
    pub not_recurring: u8,
    pub today: u8,
    pub overdue: u8,
    /// Happens now plus or minus 15min
    pub now: u8,
}

```

and these are the defaults

```rust
impl Default for SortValue {
    fn default() -> Self {
        SortValue {
            priority_none: 2,
            priority_low: 1,
            priority_medium: 3,
            priority_high: 4,
            no_due_date: 80,
            overdue: 150,
            not_recurring: 50,
            today: 100,
            now: 200,
        }
    }
}

```

It makes sense to sort by priority and due date. I also like the option of disabling our sorting and going with what Todoist returns from the API.

I would make a flag for sort, it can be one of `value`, `todoist` to start.

**Comment by alanvardy** [comment]
PR up, give it a shot and we can iterate on it easily

**Comment by stacksjb** [comment]
I think this sorting works well but is non-intuitive in that it sorts based on total weight, not in a consistent order. 

I have had tasks that meet multiple of these and figuring out how to sort consistently can be a pain.

I would prefer to sort based on a simple list-based comparator, so rather than assigning points to each item, you simply list the order you would like the sort to happen, for example:
example:
now, priority (asc), due (asc), date (asc), task (alphabetical ascending)

Basically just be able to provide a list and then it custom sorts on that.

Down the road something I can look at :)
